### ISSUE TITLE
ColorSensor driver 

### ISSUE THREAD

**Issue by to92me** [bstaletic/eurobot2k17#6]
propose API and implement driver.

**Comment by bstaletic** [comment]
@darko31 What's the exact sensor we're supposed to be using here?

It requires one timer and two GPIO pins. Do those two pins need to be external interrupts?

**Comment by darko31** [comment]
The sensor which will be used is TCS3200

http://www.mouser.com/catalog/specsheets/TCS3200-E11.pdf

Yes two pins configured as outputs for selecting color, and one pin for counting input pulses, no need for interrupts.

**Comment by bstaletic** [comment]
Maybe this driver could contain only `void read_sensor(cahr *value)` and leave everything else to the executor?

**Comment by to92me** [comment]
>Maybe this driver could contain only void read_sensor(cahr *value) and leave everything else to the executor?

or 
``` ColorEnum read_sensor(SensorStruct _ss); ``` 
or
``` void read_sensor(SensorStruct _ss, ColorEnum _ce); ```

where ```SensorStruct``` contains data about specific sensor ( port ... ) and ```ColorEnum``` si simple enum of colors.  

@bstaletic

**Comment by bstaletic** [comment]
> where SensorStruct contains data about specific sensor ( port ... ) 

If I'm getting this right the structure would have things like timer and timer channel?
I'm not sure how useful that is as we only have one such sensor and we would know up front how exactly it is connected.

**Comment by to92me** [comment]
@bstaletic We have just one color sensor ? I didn't know that :D  Ok, then there is no need for struct.

**Comment by bstaletic** [comment]
Yeah, we only need one and it has frequency as its output.

As for the API, I like the idea of returning a value from a driver. Maybe we should do that for all drivers somehow? Though I'm not sure how would we do that for analog servos.

So should it be
```c
void read_sensor(char* value);
```
In this case we don't return anything, but use pointers.

We could also use something like
```c
char* read_sensor(void);
```
Or even something different. Any more suggestions? Thoughts?

**Comment by to92me** [comment]
From point of encapsulation and abstraction that is color driver so I think that color driver needs to return a color not some value which then I have to interpret to know which color is read. 

>As for the API, I like the idea of returning a value from a driver. Maybe we should do that for all drivers somehow? Though I'm not sure how would we do that for analog servos. 

where is only one return value ( primitive built-in type or struct ) it should be return value. Where is more than one, well that depends of usage and context.

**Comment by bstaletic** [comment]
>color driver needs to return a color not some value which then I have to interpret

Fair point. Then I think it would be okay to do it like this:
```c
colour_enum_t read_sensor(void);
```
With colour_enum_t being defined like this:
```c
typedef enum { BLUE, YELLOW } colour_enum_t
```

**Comment by ivar94** [comment]
I also was thinking that driver should return value, and for example 0 if its on any color we don't need and 1 if its color we want, or something like that... but it is for the best, and its more defined for it to return color, @to92me made a pro explanation, coz he a pro :D  
@bstaletic Maybe we should make it 
```typedef enum { BLUE, YELLOW, WHITE }```
so we know that tube is not positioned on either colors?

**Comment by bstaletic** [comment]
@iVar94 
If thte tube is not positioned we shouldn't be reading the sensor in the first place.
Also, the "gibberish" that could be returned when reading in an inappropriate moment can't be guaranteed to be white. In that case the returned value could vary greatly depending on abient light.

On the other hand, everything is pretty much completely enclosed inside the robot, so not much light can enter and thus not much light can interfere. Maybe we could return a third value if we read some "gibberish".

I still would be careful about naming it a proper colour as it could be mistaken for a proper value instead of it being obvious that it is a misread value or misused sensor.

Having all that in mind how about:
```c
typedef enum { BLUE, YELLOW, ERROR } colour_enum_t;
```

**Comment by ivar94** [comment]
@bstaletic I was referring to tubes that have 4 (3 actually) colors, one blue and yellow field and two white fields, but it could work with `typedef enum { BLUE, YELLOW, ERROR } colour_enum_t;`, we just might need that info that tube is not on either color.

**Comment by darko31** [comment]
The mechanism is supposed to work as follows: spin the tube until the desired color is reached. So basically blue, yellow and whatever else, error, white, etc. 

@to92me discussed do we really need to differentiate between white and error, and we agreed that's isn't really necessary.  Just spin until the right color.

**Comment by darko31** [comment]
Solved by #26
